### PR TITLE
Freshen Makefile and update README.md 

### PR DESCRIPTION
**Pull request by virgilwashere**
# Freshen Makefile and update README.md

## README.md

- [x] Add `sudo` to revelvant commands
- [x] Use named languages (bash) in fenced codeblocks
- [x] Use `console` in fenced codeblocks for command output

### `install` target

- [x] Use `install` with permissions set

The permissions on `has` were 777 from `git clone`.

- [x] Add option to use $PREFIX

What if I don't want to install to /usr/local/bin?

`make PREFIX=$HOME/.local install` now works.

### `update` target

- [x] Add `update` target for `git fetch`
- [x] Include .PHONY targets

## tests

- [x] Add tests for Makefile changes
- [x] Uses `bats` variables for directories

- ✓ make install creates a valid installation
- ✓ ..even if has is missing from directory
- ✓ make update runs git fetch

### CODE REVIEW

**Comment by kdabir** [comment]
Thanks for the PR. To be able to merge it should pass on Travis CI. Could you please check that?

I am away right now but will certainly take a look at it little later in the week.

**Comment by virgilwashere** [comment]
> To be able to merge it should pass on Travis CI. Could you please check that?

It was and still is, passing on my local checkout.  

I will add some bats' feedback and commit shortly.

**Comment by virgilwashere** [comment]
Well.  That was a bit more effort than I expected.  But I'm ready to start using `bats` properly now.  (with `bats-assert-1` thankyou!)

File: .travis.yml [hunk 1]
```diff
@@ -1,9 +1,10 @@
 language: bash
 
 before_install:
-  - sudo add-apt-repository ppa:duggan/bats --yes
-  - sudo apt-get update -qq
-  - sudo apt-get install -qq bats
+  - git clone https://github.com/bats-core/bats-core.git /tmp/bats-core
```
Comment (virgilwashere) [inline .travis.yml]: Changed to `bats-core` `v1.1.0` here.

File: Makefile [hunk 2]
```diff
@@ -1,8 +1,32 @@
...
+test : has
 	bats .hastest.bats
 
-install:
-	cp has /usr/local/bin/has
+has :
+	git checkout --force -- has
+
+install : has
+	install -vD -m 0755 has $(DESTDIR)$(PREFIX)/bin/has
```
[the remark this answers is not in the document]
Comment (kdabir) [reply Makefile]: `$ make install` gives following error to me on on MacOS, any suitable replacement recommended?

```
install -vD -m 0755 has /usr/local/bin/has
install: illegal option -- D
usage: install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 file2
       install [-bCcpSsv] [-B suffix] [-f flags] [-g group] [-m mode]
               [-o owner] file1 ... fileN directory
       install -d [-v] [-g group] [-m mode] [-o owner] directory ...
make: *** [install] Error 64
```

[the remark this answers is not in the document]
Comment (kdabir) [reply Makefile]: By the way, Thanks a ton for taking all the effort to make tests stay green.

[the remark this answers is not in the document]
Comment (virgilwashere) [reply Makefile]: > some comment about what it does would help reader understand

My bad.  I'm normally quite pedantic about not using short options. Especially in anything going public.

`man install` on Ubuntu.
```
-d, --directory
              treat all arguments as directory names; create all components of the specified directories
-D     create all leading components of DEST except the last, or all components of --target-directory, then copy SOURCE to DEST
```

It's a shorter version of these two commands.

```
install --verbose --directory $(DESTDIR)$(PREFIX)/bin
install --verbose --mode=0755 has $(DESTDIR)$(PREFIX)/bin/has
```

I'll commit a quick patch.

> By the way, Thanks a ton for taking all the effort to make tests stay green.

I learnt how to drive `bats`.  It's time well spent.

[the remark this answers is not in the document]
Comment (kdabir) [reply Makefile]: > It's a shorter version of these two commands.
I see, however longer versions are not available on MacOs `install`'s man-page [here:](https://ss64.com/osx/install.html). closest possible is `-d` upto the bin dir

[the remark this answers is not in the document]
Comment (virgilwashere) [reply Makefile]: Sorted. 
```
install : has
        # install 'has' in specified directory
        chmod 755 has && \
        mkdir --verbose --parents $(DESTDIR)$(PREFIX)/bin && \
        cp --verbose --update has $(DESTDIR)$(PREFIX)/bin/has
```

**Comment by kdabir** [comment]
Sadly, some long opts dont work on macOS. had to change them to short flags :(
also `cp -u` is also not available. so had to remove it. :(

We can work on those in future commits. Didn't want you to fix them, so committed changes to master. you can review them if that works for you on your setup.

**Pull request merged**